server/app/services/applications/ApplicationService.java

```diff
@@ -0,0 +1,54 @@
+package services.applications;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import javax.inject.Inject;
+import models.Application;
+import play.libs.concurrent.HttpExecutionContext;
+import repository.ApplicationRepository;
+import services.program.ProgramDefinition;
+
+/** The service responsible for mediating access to the Application resource. */
+public final class ApplicationService {
+
+  private final ApplicationRepository applicationRepository;
+  private final HttpExecutionContext httpExecutionContext;
+
+  @Inject
+  ApplicationService(
+      ApplicationRepository applicationRepository, HttpExecutionContext httpExecutionContext) {
+    this.applicationRepository = checkNotNull(applicationRepository);
+    this.httpExecutionContext = checkNotNull(httpExecutionContext);
+  }
+
+  public Optional<Application> getApplication(
```

Comment:
**suggestion (blocking):** This method is named `getApplication` but its implementation is limited to checking if the application is associated with the program passed in. This is validation logic that should instead be performed wherever this check is necessary.

If you would like a convenience method for retrieving the program name of an application I recommend adding it to `Application` e.g.

```java
  public String getProgramName() {
    return getProgram().getProgramDefinition().adminName();
  }
```

Reply:
Renamed method to `validateApplication`.

Both the applicant scenario and the admin client scenario have a check for if the application has the right program name, so I believe it should stay in the service.